router.go

```diff
@@ -701,69 +693,62 @@ func (p *ControllerRegistor) ServeHTTP(rw http.ResponseWriter, r *http.Request)
 
 	}
 
-	//if no matches to url, throw a not found exception
-	if !findrouter {
+	// a "not found" exception is thrown in case no match was found
+	if !matchFound {
 		exception("404", context)
 		goto Admin
 	}
 
-	if findrouter {
-		//execute middleware filters
+	// If matchFound is true then it holds that:
+	// (routerInfo != nil || (runRouter != nil && runMethod != ""))
+	if matchFound {
+		// execute middleware filters
 		if do_filter(BeforeExec) {
 			goto Admin
 		}
-		isRunable := false
+
+		// routerInfo is non-nil only if context.Input.RunMethod and context.Input.RunMethod are nil.
+		// Therefore runMethod and runController have not been set yet.
 		if routerInfo != nil {
-			if routerInfo.routerType == routerTypeRESTFul {
+			switch routerInfo.routerType {
+			case routerTypeRESTFul:
 				if _, ok := routerInfo.methods[r.Method]; ok {
-					isRunable = true
 					routerInfo.runfunction(context)
 				} else {
 					exception("405", context)
 					goto Admin
 				}
-			} else if routerInfo.routerType == routerTypeHandler {
-				isRunable = true
+
+			case routerTypeHandler:
 				routerInfo.handler.ServeHTTP(rw, r)
-			} else {
-				runrouter = routerInfo.controllerType
-				method := r.Method
-				if r.Method == "POST" && context.Input.Query("_method") == "PUT" {
-					method = "PUT"
-				}
-				if r.Method == "POST" && context.Input.Query("_method") == "DELETE" {
-					method = "DELETE"
-				}
-				if m, ok := routerInfo.methods[method]; ok {
-					runMethod = m
-				} else if m, ok = routerInfo.methods["*"]; ok {
-					runMethod = m
-				} else {
-					runMethod = method
-				}
+
+			default:
+				runRouter = routerInfo.controllerType
+				runMethod = determineRouterMethod(routerInfo, context)
 			}
 		}
 
-		// also defined runrouter & runMethod from filter
-		if !isRunable {
-			//Invoke the request handler
-			vc := reflect.New(runrouter)
+		// internal assertion to catch bugs
+		if (runRouter != nil && runMethod == "") || (runRouter == nil && runMethod != "") {
```

Comment:
I don't get panic for `routerTypeRESTFul` and `routerTypeHandler`. They seem to work fine in the tests.

`isRunable` was set to true if and only if `runMethod` and `runRouter` were set to a non-nil value, so `isRunable` and `runMethod != "" && runRouter != nil` are equivalent.
(The only exception is the case where `context.RunMethod` and `context.RunController` are set, which was not handled by the old code).

To avoid a misunderstanding: The line `if isRunable { ...` is replaced by [line 737](https://github.com/astaxie/beego/pull/1141/files#diff-56c3bb2da6c5ecf6872e89a554944453R737) in the new code. Git's diff is confused by the additional assertion that I have inserted.